router/internal/eventorder/eventorder.go

```diff
@@ -0,0 +1,248 @@
+package eventorder
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+)
+
+var ErrUnsupportedState = errors.New("unsupported state")
+
+// NewBarrier creates a new properly initialized Barrier
+func NewBarrier(abortConcurrencyLimit int) *Barrier {
+	return &Barrier{
+		concurrencyLimit: abortConcurrencyLimit,
+		barriers:         make(map[string]*barrierInfo),
+	}
+}
+
+type Barrier struct {
+	concurrencyLimit int
+	mu               sync.RWMutex
+	queue            []command
+	barriers         map[string]*barrierInfo
+}
+
+// Enter the barrier for this userID and jobID. If there is not already a barrier for this userID
+// returns true, otherwise false along with the previous failed jobID if this is the cause of the barrier.
+// Another scenario where a barrier might exist for a user is when the previous job has failed in an unrecoverable manner and the concurrency limiter is enabled.
+func (b *Barrier) Enter(userID string, jobID int64) (accepted bool, previousFailedJobID *int64) {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
```

Comment:
Why not releasing this lock sooner? Is this done on purpose because you want to finish the operations below anyway before anybody can write in `b.barriers`?

Reply:
correct, even though `Enter` & `Sync` methods are not going to be called by multiple threads, they are synchronisation points

Reply:
Sounds like a good candidate for a comment then :wink:

Reply:
actually, access to all barrier's methods (commands) should be synchronised, thus added a relevant comment in the mutex's declaration